ExcelWriter.py

```diff
@@ -1,130 +1,228 @@
+    # Figure out the width of all the columns, sort
+    # them, then chop off the last 3 to kill off the giant
+    # outliers (3 was an arbitrary number that seemed to work)
+    # averge didn't work well, and neither did median (when we
+    # have a lot of missing assignment, the median goes to zero).
+    def Compute_Col_Width(self):
+        maxCol = []
+        colWidths=[[] for i in range(0, len(self.fontNames))]
+        for studentValues in self.entries:
+            currentCol = 0            
+            for entryList in studentValues:
+                for entry in entryList:              
+                    colWidths[currentCol].append(len(entry) + 1)
+                currentCol = currentCol + 1
         
-    def Close(self):
-        #self.Merge_Cells()
-        
+        for colWidthList in colWidths:
+            colWidthList.sort();
+            #print(colWidthList)
+            # not quite the max, trim off the last 3 entries, because some of those
+            # are very large when there is an error 
+            pos = len(colWidthList) - 3
+            if (pos < 0):
+                pos = len(colWidthList) - 1
+            maxCol.append(colWidthList[pos]);        
+        return maxCol
+    
+    # Set the widths of the columns in the spreadsheet based on the
+    # size of the strings
+    def Set_Col_Widths(self):
+        maxCol = self.Compute_Col_Width()
+        #print(maxCol)
         colNum = 0
-        for colWidth in self.maxCol:
+        for colWidth in maxCol:
             colString = self.letters[colNum] + ":" + self.letters[colNum]
             #print ("setting width " + colString + " " + str(colWidth))
             self.worksheet.set_column(colString, colWidth)
-            colNum = colNum + 1        
+            colNum = colNum + 1
+            
+    #Takes in all the lists associated with a single student
+    #figures out how many rows the entries for this student will
+    #occupy.  Sets the bottom row to have a default visible border
+    #at the bottom - this extends the ____ all the way across the 
+    #sheet.
+    def Add_Default_Border(self, studentValues, startRow, isHeader):
+        #figure out which list in this is the longest, that is how
+        #many rows this entry will occupy
+        numRows = 0
+        for entryList in studentValues:
+            numRows = max(numRows, len(entryList))            
+        endRow = numRows + startRow - 1            
+        defaultBottom = self.Get_Border_Format(0, 1, self.fontNames[0], isHeader)
+        if (endRow == startRow):
+            defaultBottom = self.Get_Border_Format(1, 1, self.fontNames[0], isHeader)
+        self.worksheet.set_row(endRow, None, defaultBottom)
+        return (numRows, endRow)
+    
+    # get the font name for printing this column.  We should use Courier New for things
+    # like the source so that indenting is not weird
+    def Get_Font_Name(self, currentCol, isHeader):
```

Comment:
I don't know if you find one-liners more or less readable, but this could be `return 'Arial' if isHeader else self.fontNames[currentCol]`.

Reply:
I sort of like the spread out flow.  I think this is going to change a few more times.